clang/utils/TableGen/ClangDiagnosticsEmitter.cpp

```diff
@@ -185,10 +174,8 @@ static void groupDiagnostics(ArrayRef<const Record *> Diags,
   }
 
   // Assign unique ID numbers to the groups.
-  unsigned IDNo = 0;
-  for (std::map<std::string, GroupInfo>::iterator
-       I = DiagsInGroup.begin(), E = DiagsInGroup.end(); I != E; ++I, ++IDNo)
-    I->second.IDNo = IDNo;
+  for (auto [IdNo, Iter] : enumerate(DiagsInGroup))
+    Iter.second.IDNo = IdNo;
```

Comment:
If you are taking each element by value (as as opposed to reference), does `Iter.second.IDNo = Idno;` actually update the original?  Or does it just update the copy?

Reply:
As far as I can tell, `llvm::enumerate` captures its ranges as references, so this might be ok?  So the effective `type` of `Iter` should be a reference to the original.

This is worth confirming however!

Either way, making the structured binding a reference wouldn't help, as that would just make the `zippy` pair be a reference, which already has 'captured' its value.

Reply:
Right, this updates the original and not the copy. See the comment at the top of enumerate() which says the same:

https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/STLExtras.h#L2411

```
/// Given two or more input ranges, returns a new range whose values are
/// tuples (A, B, C, ...), such that A is the 0-based index of the item in the
/// sequence, and B, C, ..., are the values from the original input ranges. All
/// input ranges are required to have equal lengths. Note that the returned
/// iterator allows for the values (B, C, ...) to be modified.
```

Reply:
"are the values" doesn't mean "we didn't copy these and gave you a reference".  So I don't think we can conclude my 5 minute look into the types is accurate based on that comment.

Reply:
Also look at "EnumerateModifyLValue" test in STLExtrasTest.cpp which tests this exact thing. That is, the values vended out by enumerate are references to the original values and not copies.